### PR TITLE
Fix MLX VLM training attention masking, shared K/V, embedding scale and Gemma 3n batching

### PR DESCRIPTION
**Pull request by Lyxot**
Six independent correctness bugs in MLX VLM training, each of which silently produced a wrong loss rather than an error. Four were found while building audio-input training and are extracted here so they can land on their own; the last two were found by sweeping the other families once the pattern was clear.

## The bugs

**A forwarded padding mask replaced causality.** mlx-vlm treats a supplied `mask` as the complete attention mask for every layer and only builds causal or sliding masks when it is `None`. The baseline VLM loss forwarded the collated 2-D padding mask, so causality was replaced outright and every supervised position could attend to the token it was being scored on. On `gemma-4-E2B` the loss came out at **3.9322 against a transformers reference of 7.4854** — below the reference, which is the tell: the model was reading the answer. Withholding the mask matches at 7.3659.

The fix sends `mask=None` except to the families that genuinely consume it. That is a checked list rather than a signature probe, because `molmo_point` declares `mask` and never reads it; the keyword is still always sent, because thirteen families declare it with no default; and it is matched case-insensitively on `model_type`, because a config keeps its own spelling through mlx-vlm's alias remap.

Withholding the mask leaves causality as the only thing excluding pads, and causality excludes a trailing pad, not a leading one. Collation therefore has to guarantee the layout rather than assume it. This is not a corner case: instruct checkpoints commonly ship `padding_side: left`, and five of the eleven locally exercisable families — Gemma 3, Gemma 3n, GLM-OCR, Qwen2-VL and Qwen3-VL — were producing left-padded batches before this change.

**KV-shared layers rebuilt K/V they were meant to borrow.** Gemma 3n and Gemma 4 reuse one layer's K/V in later layers, and mlx-vlm reaches that shared state only through the cache. Training forwards pass no cache, so every KV-shared layer rebuilt K/V from projections the reference implementation never uses, and the tail of the model attended to the wrong keys. `gemma-4-E2B` scored **14.3929 where transformers gives 7.4854**; supplying the shared state reaches 7.3659, matching mlx-vlm's own later implementation exactly. Gemma 3n moves from 22.1412 to 17.9880 against an 18.0012 reference and has no fixed upstream release at all.

For Gemma 4 this is a backport rather than a new fix: mlx-vlm shipped this correctly in 0.4.3, regressed it in `Fix Gemma 4 chunked prefill for KV-shared models and thinking` (#901) which first released in **0.4.4**, and repaired it in `Add KV cache quantization for continuous batching` (#1030) which first released in **0.5.0**. The affected window is exactly one release — and it is the release this package declares as its floor, so anyone on `mlx-vlm>=0.4.4` who has not upgraded is training on it. Gemma 3n is not a backport: its shared branch still reads `if self.is_kv_shared_layer and cache is not None` at 0.6.8, with no `shared_kv` threading anywhere in its language model, so there is no upstream release to move to.

No version gate is needed. On 0.5.0 and later Gemma 4 reads shared K/V from its own layer results, so the slots supplied here stop being where the sharing is read and leave the result unchanged — measured identical with and without them.

The borrowed K/V is carried as a value in the call graph rather than as an attribute on the cache. Under gradient checkpointing — on by default — an attribute is invisible to the recomputed backward pass, which silently drops the gradient while leaving the forward correct: the same model measured a gradient norm of 19.172302 against 21.723606 without checkpointing.

**Gemma 3 image training had no causality.** The multimodal path built its 4-D mask as an outer product of the padding mask, which is all-ones whenever nothing is padded. On `gemma-3-4b-it` with a 274-token image batch, **all 37,401 above-diagonal entries were visible** and the loss came out at 5.6532 against 19.0356 for the same batch through the cross-entropy path. Both paths now agree exactly, and the remaining above-diagonal visibility is 32,640 entries, which is the interior of the single 256-token image block and nothing else.

**Gemma 3n lost its embedding scale on every multimodal forward.** Its language model multiplies token embeddings by `sqrt(hidden_size)` only when given input ids; every multimodal forward hands it merged embeddings instead. The cross-entropy path scored **12.1275 where transformers gives 18.0012**, and now reaches 17.9937. Only positions still holding a plain token embedding are scaled — merged features already arrive at the scaled magnitude — and those positions are found by comparing against the token embeddings rather than by token id, because Gemma 3n's closing audio delimiter carries a feature while not being the audio token.

**PaliGemma's suffix was not causal.** Same construction as the Gemma 3 bug, in a separate implementation that the Gemma 3 fix does not reach, and still present upstream at 0.6.8 — this one has nowhere to upgrade to either. PaliGemma is a prefix-LM: its image and prompt are meant to attend bidirectionally and only the suffix is causal. Its 4-D mask was an outer product of the padding mask, so on an eight-position batch with a four-token prefix all 28 above-diagonal entries were visible and **22 of them were a completion token reading its own future**. After the fix the same batch keeps 6, exactly the prefix block's interior, with the prefix still fully bidirectional.

Both loss paths are corrected. The cross-entropy path embeds directly and already carries the token types that mark the prefix boundary; the plain path goes through `Model.__call__`, which embeds with a fixed three arguments and drops them, so a companion wrapper carries them across on a per-thread stack — not on the model, since one model can serve several callers at once and an attribute would let one read another's batch.

**Gemma 3n could not train above a batch of one.** Its AltUp correction transposes its coefficients as though the batch axis were last, giving `(inputs, sequence, batch)` against an innovation of `(batch, sequence, hidden)`. At a batch of one the trailing axis is 1 and broadcasts by accident; anything larger raises `[broadcast_shapes] Shapes (1,3,24,2048) and (4,1,24,3) cannot be broadcast`. Transposing to `(inputs, batch, sequence)` with an explicit trailing axis makes the coefficients broadcast over hidden instead of colliding with it. A ragged batch of three now trains where it previously raised, and a batch of one is bit-identical before and after.

This is the second backport of mlx-vlm #1030 — the same upstream commit that repaired Gemma 4's KV sharing, for the same reason. The replacement installs only after the loaded implementation is observed to fail a two-row correction, so a release that already handles it is left untouched and no version is hard-coded.

## Validation

Every number above is measured, not estimated. Beyond that:

**Ragged batches larger than one.** Every earlier validation of these fixes ran at batch size 1, where there is no padding at all and therefore no way to see a padding-layout defect. Collation was re-checked on real processors at ragged batch 3 across every locally available family, asserting content-then-padding on every row and that each token-aligned sidecar still marks the same tokens after a repair. The sidecar check has to compare positionally: an image block is a run of one repeated soft token, so a shift inside it leaves the multiset of marked ids identical while every mark points at the wrong column.

**No regression on other families.** Losses are bit-identical between this branch and `main` where no change is intended — `idefics3` at 6.443582 on both mlx-vlm 0.4.4 and 0.6.8, `gemma3` at 4.000300. Qwen2-VL's loss changes, and only because its layout is corrected: both sides supervise the same 88 positions and neither supervises a pad. A behavioural causality probe that perturbs a future token and checks whether an earlier logit moves reports zero drift for every family on both sides, so withholding the mask cannot make a family less causal than it was.

**A reference check on the padding argument.** Under causal-only attention a right-padded batch should reproduce the loss you get when the padding mask is supplied, and a left-padded batch should not. On the reference implementation with no MLX involved:

```text
right-padded   with padding mask 3.664971 | causal-only 3.662987 | drift 0.001984
left-padded    with padding mask 3.937409 | causal-only 3.412554 | drift 0.524856
```

Right-padded agrees to 0.002 nats; left-padded is 0.52 out and lower, the same "better than the reference" signature as the original defect.

**Suite.** No test fails on this branch that does not also fail on `main`. Note that `tests/test_moe_fused_narrow_expert_merge.py::test_fused_narrow_expert_merge_applies_delta` is nondeterministic on `main` itself — five full-suite runs there gave 167, 167, 166, 167 and 168 failures — so the comparison was made over unions of failing node ids across several runs per side rather than single runs.

## Relationship to upstream

Two of these have upstream fixes, both in mlx-vlm #1030, first released in 0.5.0: Gemma 4's shared K/V and Gemma 3n's AltUp batching. Those two parts are backports so the declared `mlx-vlm>=0.4.4` floor is correct, and both become inert once a user is on 0.5.0 or later — the AltUp one detects that for itself. Everything else is still live at 0.6.8: Gemma 3n's shared layers remain cache-tied, and the padding-outer-product mask remains in both `gemma3.py` and `paligemma.py`. The remaining two — the forwarded padding mask and the lost embedding scale — are in this package's own call sites, not upstream's.

## Risks and limits

- The Gemma 3n embedding scale is restored on the cross-entropy path, which is the trainer default, and not on the `use_cce=False` path, which calls the model's own forward and embeds internally where this cannot reach.
- The PaliGemma correction needs the processor's token types, which is where the prefix boundary lives. transformers 5.x returns those unconditionally; 4.x returns them only for a row carrying a suffix, so on 4.x a chat-style row keeps upstream's mask. The boundary is not recoverable at the embedder — only the supervised positions identify the suffix, and labels do not reach it — and guessing would be worse than leaving it, since a plain causal mask would also make the image block causal, which PaliGemma is not trained for. Where token types are absent the mask is left exactly as upstream built it, so this is a strict improvement rather than a change of behaviour.
- Sidecar relocation is driven by a named key set rather than inferred from array shapes. Shape does not identify what an array is indexed by — per-image metadata can carry the same extents as a per-token array — so an unlisted sidecar is deliberately left alone rather than guessed at, which is the same stance the width survey already takes on the same ambiguity.
- The collation repair is a safety net, not a normal path: every family checked honours `padding_side="right"`, so it only runs for processors that ignore the request, which `deepseek_vl_v2` and the Falcon pair do.
- `lfm2_vl` fails to train on mlx-vlm 0.4.4. That is an upstream issue, it reproduces identically on `main`, and it is not addressed here.

### CODE REVIEW

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/loader.py [hunk 1]
```diff
@@ -2367,6 +2369,198 @@ def prepare_inputs_for_multimodal(
...
+
+
+def _paligemma_replace_mask(features, token_type_ids, attention_mask):
+    """Swap the outer-product mask for the prefix-LM one where the split is known.
+
+    Without token types there is no prefix boundary to derive, so upstream's mask
+    is left as it is rather than guessed at.
+    """
+    if token_type_ids is None or features.attention_mask_4d is None:
+        return features
```
[the remark this answers is not in the document]
Comment (Lyxot) [reply unsloth_zoo/mlx/loader.py:2453]: Checked this against both ends of the supported `transformers` range, and it is version-dependent rather than universal — but you are right that a real gap exists, so thank you.

`PaliGemmaProcessor.__call__`:

- **transformers 5.x** — `return_token_type_ids = True`, unconditionally, with the source even noting "we always return token_type_ids". Token types are always present, so the replacement always applies.
- **transformers 4.x** — `return_token_type_ids = suffix is not None`. Token types appear only when a suffix is passed, which collation does whenever a row carries one, and not otherwise.

This package supports `>=4.51.3,<=5.5.0`, so the gap is real but narrower than stated: it is 4.x **and** a row without a suffix. On 5.x, or on 4.x with prompt/completion rows, the replacement applies normally. The `_vlm_processor_requests_mm_token_type_ids` part of the report is a separate key — that gates `mm_token_type_ids`, which this code does not read.

I have not closed the remaining gap here, deliberately:

1. **The boundary is genuinely unavailable at this point.** The only thing that identifies the suffix is which positions are supervised, and that lives in `labels`, which neither loss path forwards to the embedder. Closing it means threading a completion boundary from collation through both loss paths into this patch — a cross-file change that wants its own change set and its own review, and this one is deliberately scoped to the mask.
2. **The alternatives are guesses.** Falling back to a plain causal mask would remove the leak but also make the image block causal, which is wrong for PaliGemma and would degrade training in a way that is harder to notice than the leak. Deriving the prefix from the image layout does not work either, since the prompt continues past the last image and the suffix boundary is exactly what is missing.
3. **Note that on 4.x without a suffix the reference does not treat the row as training at all** — no suffix means no labels and `is_training` is false — so the fully-bidirectional mask is upstream's own behaviour for that shape. That does not make the leak acceptable for this package, which builds its own labels, but it does mean guessing a boundary would diverge from the reference rather than match it.

What this change does is a strict improvement with no regression: where token types exist the suffix becomes causal, and where they do not the mask is left exactly as upstream built it. I have made the precondition explicit in the commit message and the PR description instead of leaving it implied, and the code already says it at the call site.

**Comment by Lyxot** [comment]
@codex review

File: unsloth_zoo/mlx/loader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/loader.py:2453]: Marking this up, and I think the transformers 4.x framing understates it. Measured on transformers 5.5.0, where you say token types are always present.

They are always present, but they carry no boundary unless a suffix is passed. `PaliGemmaProcessor.__call__` derives them from `text_pair=suffix`, so a row with no suffix is segment 0 end to end. `_paligemma_prefix_lm_mask` then computes `prefix = (token_type_ids == 0)`, the whole row is prefix, and the resulting mask is fully bidirectional:

```
all-zero token types (no suffix column)   future positions visible: 28/28
suffix present (last 3 = segment 1)       future positions visible: 10/28
```

So on 5.x the guard you added does fire, and it installs a mask that lets every supervised position see the tokens after it. `suffixes` is threaded from one place, `utils.py:6192`, sourced from a literal `"suffix"` dataset column at `:6187`, so every messages/text row misses the boundary. For comparison, upstream `modeling_paligemma.py` refuses this outright at training time with `token_type_ids is required as a model input when training`.

Two things I want to be accurate about rather than overstate. First, I have not shown that this PR *introduces* the leak: for an all-zero row the replacement may well be equivalent to the mask that was already in place, in which case the finding is that the fix does not cover this case rather than that it regresses anything. Second, the remedy has a real limit. The boundary is derivable at the same call site for completion-only training, since `_vlm_cce_forward` holds `input_ids` and `assistant_token_id` in scope at `utils.py:2169-2171` and the embed call is at `:2194`. But `assistant_token_id` is `0` unless `train_on_completions` is set, so in the default full-sequence configuration there is no supervised-span boundary to recover and PaliGemma's prefix/suffix split is genuinely undetermined.

For the record on provenance, since it came up: the 4.x/5.x change is huggingface/transformers#40917, merged 2025-09-23, which replaced `return_token_type_ids = suffix is not None` with `return_token_type_ids = True`. Your description of both versions is accurate; it just does not make the suffix-less row safe.

**Comment by danielhanchen** [comment]
Heads up on merge order, from reviewing #966 which is stacked on this.

`_shared_kv_slot_count` arrives with this PR (it does not exist at merge-base `90926063`) and does not check whether the backbone shares K/V natively. mlx-vlm 0.5.0+ threads shared K/V itself, and `_fix_gemma4_kv_sharing` correctly leaves those backbones alone, but the slot list is still built and both loss paths forward it as `cache`. That list is shorter than the layer count by construction, since `first_kv_shared_layer_idx` is `num_hidden_layers - num_kv_shared_layers`, so zipping layers against it truncates the traversal.

Reproduced against `40b3f5b8`, this PR's head, on a native-shaped backbone:

    native shared kv         : True
    layers / cache slots     : 4 / 2
    VERDICT                  : TRUNCATES

So the shared tail of the stack is silently skipped on any current mlx-vlm. Nothing raises.

The fix is `6afaadac` on the #966 branch: detect native sharing, return no slots, leave `cache` unset so those backbones keep the ordinary per-layer contract. Legacy backbones are unchanged.

If the two land together this is already covered. If this one merges on its own first, the truncation goes into main and stays until #966 lands, so it would be worth cherry-picking `6afaadac` here first.
